docs(css): Fx117 - new pages for basic shape functions rect() and xywh() (#28565)
mdn/content: files/en-us/web/css/basic-shape/rect/index.md

---
title: rect()
slug: Web/CSS/basic-shape/rect
page-type: css-function
browser-compat: css.types.basic-shape.rect
---

{{CSSRef}}

> The **`rect()`** [CSS](/en-US/docs/Web/CSS) function creates a rectangle at the specified distance from the top and left edges of the containing block. It is a basic shape function of the {{cssxref("&lt;basic-shape&gt;")}} [data type](/en-US/docs/Web/CSS/CSS_Types). CSS properties such as {{cssxref("clip-path")}} and {{cssxref("offset-path")}} use the `rect()` function to create the rectangular shape path along which an element moves.

Author: Clip yes and [that's `clip-path` now](https://developer.mozilla.org/en-US/docs/Web/CSS/clip). I don't see any reference to `rect()` being used in `shape-outside` (either in [spec](https://drafts.csswg.org/css-shapes-1/#propdef-shape-outside) or in code or [wpt](https://github.com/web-platform-tests/wpt/tree/master/css/css-shapes/shape-outside/supported-shapes)).

You're right in pointing out that it's not just motion, so I've updated the last sentence to:

> You can use the `rect()` function in CSS properties such as {{cssxref("offset-path")}} to create the rectangular shape path along which an element moves and in {{cssxref("clip-path")}} to define the shape of the clipping region.

## Syntax

```css

> offset-path: rect(0px 1% auto 3% round 0 1px);

Author: Also added a `clip-path` example here:

> `clip-path: rect(50px 70px 80% 20%);`

```

### Values

> The inset rectangle is specified using four values, going clockwise starting with the top edge. Each value is a `<length>`, a `<percentage>`, or the keyword `auto`.

Reviewer: ```suggestion
The inset rectangle is specified by defining four offset values, starting with the top edge offset and going clockwise, and an optional `round` keyword with a `border-radius` parameter to provide rounded corners. Each offset value is a `<length>`, a `<percentage>`, or the keyword `auto`.
```

Author: Fixed to:

> The inset rectangle is defined by specifying four offset values, starting with the top edge offset and going clockwise, and an optional `round` keyword with the `border-radius` parameter to add rounded corners to the rectangle. Each offset value can be either a `<length>`, a `<percentage>`, or the keyword `auto`.

- `<length-percentage>`

>   - : Specifies the {{cssxref("&lt;length-percentage&gt;")}} value for the distance of the top, right, bottom, and left edges of the rectangle from the containing block. The first (top) and third (bottom) values are distances from the top edge of the containing block, and the second (right) and fourth (left) values are distances from the left edge of the containing block. The second (right) and third (bottom) values are limited by the fourth (left) and second (top) values, respectively, to correct the bottom edge from crossing over the top edge and right edge from crossing over the left edge. For example, `rect(10px 0 0 20px)` gets corrected to `rect(10px 20px 10px 20px)`.

Reviewer: ```suggestion
  - : Specifies the {{cssxref("&lt;length-percentage&gt;")}} value of the distance from the top, right, bottom, or left edge of the rectangle from the top or left edge of the containing block. The first (top) and third (bottom) values are distances from the top edge of the containing block, and the second (right) and fourth (left) values are distances from the left edge of the containing block. The second (right) and third (bottom) values are clamped by the fourth (left) and second (top) values, respectively, to prevent the bottom edge from crossing over the top edge and right edge from crossing over the left edge. For example, `rect(10px 0 0 20px)` is clamped to `rect(10px 20px 10px 20px)`.
```

Reviewer: clamped or floored

Author: - In the first sentence, I am going with:

> value of the distance of

- I wanted to use a different word and not "floored" as used in spec, so that everyone is clear on what the word means. I'll use "clamped" - I hope folks understand it.

>   - : Specifies the {{cssxref("&lt;length-percentage&gt;")}} value of the distance of the top, right, bottom, or left edge of the rectangle from the top or left edge of the containing block. The first (top) and third (bottom) values are distances from the top edge of the containing block, and the second (right) and fourth (left) values are distances from the left edge of the containing block. The second (right) and third (bottom) values are clamped by the fourth (left) and second (top) values, respectively, to prevent the bottom edge from crossing over the top edge and right edge from crossing over the left edge. For example, `rect(10px 0 0 20px)` is clamped to `rect(10px 20px 10px 20px)`.

Author: It was this version originally :) https://github.com/mdn/content/pull/28565#discussion_r1294224919.
I am, however, okay with keeping the "or" because at a time, only one of the edges can be specified.

- `auto`

>   - : Makes the edge for which this value is used to coincide with the corresponding edge of the containing block. If `auto` is used for the first (top) or fourth (left) value, the value of `auto` is equivalent to 0%, and if used for the the second (right) or third (bottom) value, `auto` value is equivalent to 100%.

Reviewer: ```suggestion
  - : Makes the edge for which this value is used to coincide with the corresponding edge of the containing block. If `auto` is used for the first (top) or fourth (left) value, the value of `auto` is equivalent to `0%`, and if used for the second (right) or third (bottom) value, the `auto` value is equal to `100%`.
```

> - `round <border-radius>`

Reviewer: ```suggestion
- `round <\`border-radius\`>`
```

  - : Specifies the radius of the rounded corners of the rectangle using the {{cssxref("border-radius")}} shorthand syntax. This parameter is optional.

## Examples

### Creating offset-path using rect()

In this example, the {{cssxref("offset-path")}} property uses the `rect()` function to define the shape of the path on which the element, red box in this case, moves. Three different scenarios are shown, each using different values for the `rect()` function.

```html
<div class="container">
  Rectangular path 1
  <div class="path rect-path-1"></div>
</div>
<div class="container">
  Rectangular path 2
  <div class="path rect-path-2"></div>
</div>
<div class="container">
  Rectangular path 3
  <div class="path rect-path-3"></div>
</div>
```

```css
.container {
  position: relative;
  display: inline-block;
  width: 200px;
  height: 200px;
  border: 1px solid black;
  margin: 15px;
  text-align: center;
}

.path {

>   width: 50px;
>   height: 50px;

Reviewer: ```suggestion
  width: 40px;
  height: 40px;
```

  position: absolute;
  animation: move 10s linear infinite;
}

> .rect-path-1 {
>   background-color: red;
>   offset-path: rect(50px 150px 200px 50px round 20%);
> }
> 
> .rect-path-2 {
>   background-color: red;
>   offset-path: rect(50px auto 200px 50px round 20%);
> }
> 
> .rect-path-3 {
>   background-color: red;
>   offset-path: rect(50px auto 200px auto);
> }

Reviewer: ```suggestion
.rect-path-1 {
  background-color: red;
  offset-path: rect(30px 120px 180px 30px round 20%);
}

.rect-path-2 {
  background-color: red;
  offset-path: rect(30px auto 180px 30px round 20%);
}

.rect-path-3 {
  background-color: red;
  offset-path: rect(30px auto 180px auto);
}
```

Reviewer: these changes are so all 3 can go on one line. Alternative, make the example taller

Author: I've reduced the size of the red box though from 50px x 50px to 40px x 40px.

@keyframes move {
  0% {
    offset-distance: 0%;
  }
  100% {
    offset-distance: 100%;
  }
}
```

#### Result

> {{EmbedLiveSample("Creating an offset-path using rect", "100%", 300)}}

Reviewer: either make taller or go with my edits abov e

Author: I've increased the height of the frame.

> - Path 1 rectangle specifies distances of all edges (top, right, bottom, and left) from the containing block. Top and bottom values are distances from the top edge of the containing block. Right and left values are distances from the left edge of the containing block. In addition, the corner of the rectangle is rounded`20%`, so the element moving on this path follows the rounded corner.

Reviewer: ```suggestion
- The rectangle in Path 1 specifies the distances of the four edges (top, right, bottom, and left) from the containing block. The top and bottom values are distances from the top edge of the containing block. The right and left values are distances from the left edge of the containing block. In addition, the corner of the rectangle is rounded at `20%`, making the element follows the rounded corners as it moves along this path.
```

Author: Path 1 is the rectangle itself, so not using "the rectangle in Path 1". I've updated the text to:

> - The path 1 rectangle specifies the distances of the four edges (top, right, bottom, and left) from the containing block. The top and bottom values are distances from the top edge of the containing block. The right and left values are distances from the left edge of the containing block. In addition, the corner of the rectangle is rounded at `20%`, making the red box element follow the rounded corners as it moves along this path.

> - Path 2 is similar to path 1 except that the right value is `auto`. As a result, the right edge of the rectangle matches the right edge of the containing block.

Reviewer: ```suggestion
- Path 2 is similar to path 1, except that the right value is `auto`, which is the equivalent of 100%, causing the right edge of the rectangle to match the right edge of the containing block, creating a wider rectangle than path 1.
```

Author: Updated to:

> - The path 2 rectangle is similar to the path 1 rectangle, except that the right value is `auto`, which is equal to the value `100%`. This causes the right edge of the rectangle to match the right edge of the containing block, creating a wider rectangle than path 1.

> - Path 3 omits the corner rounding. Also, both right and left values are set to `auto`.

Reviewer: ```suggestion
- Path 3 has both the left and right edge parameters set to `auto` and omits the `round <\`border-radius\`>` parameter creating a rectangle that is the width of the reference box, with rectangular corners, instead of rounded corners like the other two examples.
```

Author: Updated to:

> - The path 3 rectangle specifies both the left and right edge parameters as `auto` and omits the `round <border-radius>` parameter. This creates a rectangle that has the width of the containing block and rectangular corners instead of rounded corners like in path 1 and path 2 rectangles.

## Syntax

```css
offset-path: rect(0 1% auto 3% round 0 1px);
clip-path: rect(50px 70px 80% 20%);
```

### Values

The inset rectangle is defined by specifying four offset values, starting with the top edge offset and going clockwise, and an optional `round` keyword with the `border-radius` parameter to add rounded corners to the rectangle. Each offset value can be either a `<length>`, a `<percentage>`, or the keyword `auto`.

- `<length-percentage>`

  - : Specifies the {{cssxref("&lt;length-percentage&gt;")}} value of the distance of the top, right, bottom, or left edge of the rectangle from the top or left edge of the containing block. The first (top) and third (bottom) values are distances from the top edge of the containing block, and the second (right) and fourth (left) values are distances from the left edge of the containing block. The second (right) and third (bottom) values are clamped by the fourth (left) and second (top) values, respectively, to prevent the bottom edge from crossing over the top edge and right edge from crossing over the left edge. For example, `rect(10px 0 0 20px)` is clamped to `rect(10px 20px 10px 20px)`.

- `auto`

>   - : Makes the edge for which this value is used to coincide with the corresponding edge of the containing block. If `auto` is used for the first (top) or fourth (left) value, the value of `auto` is `0%`, and if used for the second (right) or third (bottom) value, the value of `auto` is `100%`.

Author: Thanks
